app/src/services/cpk.ts

```diff
@@ -1473,20 +1476,90 @@ class CPKService {
-      txOptions.gas = defaultGas
+      // setTxHash: (arg0: string) => void
+      //     setTxState: (step: TransactionStep) => void
+      const { chainId } = await this.provider.getNetwork()
+      const OmenGuild = new OmenGuildService(this.provider, chainId)
+
+      if (chainId !== networkIds.MAINNET || this.cpk.relay) {
+        const signer = this.provider.getSigner()
+        const account = await signer.getAddress()
+
+        const omenTokenAddress = await OmenGuild.omenTokenAddress()
+        const allowanceAddress = await OmenGuild.tokenVault()
+        const collateralService = new ERC20Service(this.provider, account, omenTokenAddress)
+        const transactions: Transaction[] = []
 
-      const transactions: Transaction[] = [
-        {
-          to: tokenAddress,
-          data: ERC20Service.encodeApproveUnlimited(OMNI_BRIDGE_XDAI_ADDRESS),
-        },
-      ]
-      return this.execTransactions(transactions)
+        const txOptions: TxOptions = {}
+
+        const hasCPKEnoughAlowance = await collateralService.hasEnoughAllowance(
+          this.cpk.address,
+          allowanceAddress,
+          amount,
+        )
+
+        if (!hasCPKEnoughAlowance) {
+          // Step 1:  Approve unlimited amount to be transferred to the token vault if not approved already
+          transactions.push({
+            to: omenTokenAddress,
+            data: ERC20Service.encodeApproveUnlimited(allowanceAddress),
+          })
+        }
+        // Step 2: Fund CPK
+        transactions.push({
+          to: omenTokenAddress,
+          data: ERC20Service.encodeTransferFrom(account, this.cpk.address, amount),
+        })
+        // Step 3: Lock Tokens in the guild
+        transactions.push({
+          to: OmenGuild.omenGuildAddress,
+          data: OmenGuildService.encodeLockTokens(amount),
+        })
+
+        const { transactionHash } = await this.execTransactions(transactions, txOptions, setTxHash, setTxState)
+        return transactionHash
+      } else {
+        const transaction = await OmenGuild.lockTokens(amount)
+
+        return transaction.hash
+      }
+    } catch (e) {
+      logger.error(`Error while trying to lock Omen tokens : `, e.message)
+      throw e
+    }
+  }
+  unlockTokens = async (amount: BigNumber) => {
+    try {
+      const network = await this.provider.getNetwork()
+      const OmenGuild = new OmenGuildService(this.provider, network.chainId)
+      if (this.cpk.relay) {
+        const txOptions: TxOptions = {}
+        txOptions.gas = defaultGas
+
+        const transactions: Transaction[] = [
+          {
+            to: OmenGuild.omenGuildAddress,
+            data: OmenGuildService.encodeUnlockTokens(amount),
+          },
+        ]
+
+        const { transactionHash } = await this.execTransactions(transactions, txOptions)
+
+        return transactionHash
+      } else {
+        const transaction: Transaction = await OmenGuild.unlockTokens(amount)
+
+        return transaction
+      }
```

Comment:
Should there also be an if `!mainnet && !relay`, unlock tokens via cpk and transfer to user address?